039-var-sources/proposal.md

```diff
@@ -0,0 +1,187 @@
+shared across teams within an installation. This means not relying on cluster
+state for access control; resource types should entirely trust their `source:`.
+
+This is problematic for resources which make use of IAM roles associated to
+their `worker` EC2 instances in order to authenticate, because in this case the
+resource's `source:` does not actually include any credentials. As a result, we
+cannot safely enable [global resources][global-resources-opt-out]. by default
+because these resources would share version history without even vetting their
+credentials.
+
+A special credential manager could be implemented to acquire credentials via
+IAM roles on the `web` EC2 instance and then provide them to the `source:`
+configuration via `((vars))`. This way the `source:` configuration is the
+source of truth. This is discussed in [concourse/concourse#3023][issue-3023].
+
+However, as there can only be one credential manager configured at a time,
+using that single "slot" just for IAM roles is a bit of a waste compared to a
+full-fledged credential manager that can be used for many more things.
+
+# Proposal
+
+This proposal introduces a new kind of configuration: `var_sources`.
+
+This name "var source" is chosen to build on the existing terminology around
+`((vars))` and to directly relate them to one another. Calling them "var
+sources" instead of "credential managers" also allows them to be used for
+things that aren't necessarily credentials.
+
+`var_sources` may be specified at a pipeline-level, like so:
+
+```yaml
+var_sources:
+- name: vault
+  type: vault
+  config:
+    uri: https://vault.example.com
+    # ... vault-specific config including auth/etc ...
+
+resources: # ...
+
+jobs: # ...
+```
+
+Each var source has a `name` which must be a [valid
+identifier][valid-identifier-rfc]. This is used to explicitly reference the
+source from `((vars))` syntax so that there is no ambiguity. See
+[`VAR_SOURCE_NAME`](#VAR_SOURCE_NAME).
+
+A var source's `type` specifies one of the supported credential managers, e.g.
+`vault`, `credhub`, `kubernetes`. is responsible for interpreting
+`config`.
+
+A var source's `config` is a "black box" to Concourse and is passed verbatim to
+the credential manager. This configuration should include any credentials
+necessary for authenticating with the credential manager.
+
+A var source's `config` may use `((vars))` to obtain its own credentials,
+either by using a different var source or by using the system-level credential
+manager. These may also be provided statically at pipeline-configuration time.
+Either way, the config will be encrypted when stored in the database.
+
+## `((var))` syntax
+
+The `((var))` syntax was introduced a long while back and was never formally
+specified or documented. This RFC proposes a change to it so now's a good time
+to lay it all out.
+
+The full `((var))` syntax will be
+**`((VAR_SOURCE_NAME:SECRET_PATH.SECRET_FIELD))`**.
```

Comment:
Do we want to support multiple var source names? Like: `((vault1,vault2:password))`. This might add some flexibility.

And think we may still support `((secret))` to iterate all var sources.

Reply:
Personally I don't see a need for specifying multiple var source names. Seems like the user should know where their secrets are coming from if they know enough to suggest a couple of specific sources. :thinking:

Querying multiple var sources could result in surprises. For example, if their pipeline relied on a credential `foo` from `vault2` but then a credential under the same name was added to `vault1`.

Reply:
Make sense.